File: src/main/resources/static/js/data-rope.js
var ropeRoot =  new TreeNode(null, null, null,null, null);

const usedStyles = new Map();

let specChars = 0;

const STYLE_CODES = {
    'BOLD' : 0,
    'ITALIC' : 1,
    'UNDERLINE' : 2,
    'STRIKETHROUGH' : 3,
    'FONT' : 4,
    'FONT_COLOUR' : 5,
    'FONT_SIZE' : 6,
    'LIST_ORDERED' : 7,
    'LIST_UNORDERED' : 8,
    'HEADER' : 9
}

function displayTree() {
    if (document.getElementById) {
        w = screen.availWidth;
        h = screen.availHeight;
    }
    let popW = 800, popH = 700;
    let leftPos = (w-popW)/2;
    let topPos = (h-popH)/2;
    let msgWindow = window.open('/static/treant-display.html',
        'popup',
        'width=' + popW + ',height=' + popH + ',top=' + topPos + ',left=' + leftPos + ',scrollbars=yes');
}

function TreeNode(parent, left, right, text, style) {
    this.parent = parent;
    this.children = [left, right];
    this.text = text;
    this.length = text ? text.length : null;
    this.style = style;
    //functions
    this.getText = () => {
        return text;
    }
    this.getLength = () => {
        return this.text ? this.text.length : this.length;
    }
    this.getLeft = () => {
        return this.children[0];
    }
    this.getRight = () => {
        return this.children[1];
    }
    this.setText = (text) => {
        this.text = text;
        this.updateLength();
    }
    this.setLeft = (left) => {
        this.children[0] = left;
        left.parent = this;
        this.updateLength();
    }
    this.setRight = (right) => {
        this.children[1] = right;
        right.parent = this;
        this.updateLength();
    }
    this.updateLength = () => {
        if (this.text == null)
            this.length = (this.getLeft() ? this.getLeft().getLength() : 0) + (this.getRight() ? this.getRight().getLength() : 0);
        else
            this.length = this.text.length;
        if (this.parent)
            this.parent.updateLength();
    }
    this.deleteLeft = () => {
        if (this.getLeft()) {
            this.getLeft().deleteLeft();
            this.getLeft().deleteRight();
            this.children[0].text = null;
            this.children[0] = null;
        }
        this.updateLength();
    }
    this.deleteRight = () => {
        if (this.getRight()) {
            this.getRight().deleteLeft();
            this.getRight().deleteRight();
            this.children[1].text = null;
            this.children[1] = null;
        }
        this.updateLength();
    }
    this.isLeft = () => {
        return this.parent.getLeft() === this;
    }
    this.nextTextNode = () => {
        if (this.text) return nextTextNode(this);
        else return leftmostChild(this);
    }
    this.prevTextNode = () => {
        if (this.text) return prevTextNode(this);
        else return leftmostChild(this);
    }
    this.getOffset = () => {
        return getOffset(this);
    }
}

function leftmostChild(node){
    let el = node;
    while (el.text == null){
        if (el.getLeft())
            el = el.getLeft();
        else
            el = el.getRight();
    }
    return el;
}

function ropeInsertText(text, style, pos) {
    if (text.match(/[\n]/g))
        specChars++;

    let newNode = new TreeNode(null, null, null, text, style);

    let added = [];
    let removed = [];
    if (pos === 0 || pos === ropeRoot.length){
        if (ropeRoot.getLeft() || ropeRoot.getRight() || ropeRoot.getText()) {
            if (pos === 0){
                if (ropeRoot.getLeft() == null)
                    ropeRoot.setLeft(newNode);
                else if (ropeRoot.getRight() == null){
                    ropeRoot.setRight(ropeRoot.getLeft());
                    ropeRoot.setLeft(newNode);
                }
                else
                    ropeRoot = concat(newNode, ropeRoot);
            }
            else {
                if (pos === ropeRoot.getLength()) {
                    if (ropeRoot.getRight() == null)
                        ropeRoot.setRight(newNode);
                    else if (ropeRoot.getLeft() == null) {
                        ropeRoot.setLeft(ropeRoot.getRight());
                        ropeRoot.setRight(newNode);
                    } else
                        ropeRoot = concat(ropeRoot, newNode);
                }
            }
        }
        else
            ropeRoot.setRight(newNode);

        return {added : [newNode], removed};
    }

    let toChange = getAffectedNode(ropeRoot, pos);
    let affectedNode = toChange[0];
    let remainingPos = pos - toChange[1];

    let isLeft = affectedNode.parent.getLeft() === affectedNode;

    if (remainingPos === 0) {
        if (isLeft) {
            if (affectedNode.parent.getRight() == null) {
                affectedNode.parent.setRight(affectedNode);
                affectedNode.parent.setLeft(newNode);
            } else {
                affectedNode.parent.setLeft(concat(newNode, affectedNode));
            }
        } else {
            if (affectedNode.parent.getLeft() == null) {
                affectedNode.parent.setLeft(newNode);
            } else {
                affectedNode.parent.setRight(concat(newNode, affectedNode));
            }
        }
        added.push(newNode);
    } else if (remainingPos === affectedNode.length) {
        if (isLeft) {
            if (affectedNode.parent.getRight() == null) {
                affectedNode.parent.setRight(newNode);
            } else {
                affectedNode.parent.setLeft(concat(affectedNode, newNode));
            }
        } else {
            if (affectedNode.parent.getLeft() == null) {
                affectedNode.parent.setLeft(affectedNode);
                affectedNode.parent.setRight(newNode);
            } else {
                affectedNode.parent.setRight(concat(affectedNode, newNode));
            }
        }
        added.push(newNode);
    } else {
        let [first, second] = splitString(affectedNode.text, remainingPos);
        let node1 = new TreeNode(null, null, null, first, affectedNode.style);
        added.push(node1);
        node1 = concat(node1, newNode);
        let node2 = new TreeNode(null, null, null, second, affectedNode.style);
        node1 = concat(node1, node2);
        if (isLeft)
            affectedNode.parent.setLeft(node1);
        else
            affectedNode.parent.setRight(node1);
        added.push(newNode);
        added.push(node2);
        removed.push(affectedNode);
    }

    return {added, removed};
}

function ropeDeleteText(pos, length) {
    let start = getAffectedNode(ropeRoot, pos + 1)
    let affected = start[0];
    let remainingPos = pos - start[1];
    if (remainingPos < 0)
        remainingPos = ropeRoot.getLength() -Math.abs(pos - start[1]);

    let remainingLength = length;

    let lastNode = affected;
    let nodeLength = lastNode.getLength();
    let parent = lastNode.parent;
    let isLeft = lastNode.isLeft();

    let changed = [];
    if (length === ropeRoot.getLength()){
        let el = leftmostChild(ropeRoot);
        while (el){
            changed.push({el : el, before : el.text});
            if (el.text.match(/[\n]/g))
                specChars--;

            el = el.nextTextNode();
        }
        ropeRoot.deleteLeft();
        ropeRoot.deleteRight();
        return changed;
    }

    if (remainingPos > 0) {
        let toDelete = clamp(remainingLength, 0, nodeLength);
        remainingLength -= toDelete;
        changed.push({el : lastNode, before : lastNode.text});
        if (lastNode.text.match(/[\n]/g))
            specChars--;
        deletePartFromTextNode(lastNode, remainingPos, toDelete);
        lastNode = nextTextNode(lastNode);
    }
    else {
        changed.push({el : lastNode, before : lastNode.text});
        if (lastNode.text.match(/[\n]/g))
            specChars--;
        if (nodeLength > remainingLength) {
            deletePartFromTextNode(lastNode, 0, remainingLength);
            remainingLength = 0;
        }
        else {
            lastNode.text = null;
            remainingLength -= nodeLength;
            lastNode = nextTextNode(lastNode);
            isLeft ? parent.deleteLeft() : parent.deleteRight();
            while (parent.getLeft() == null && parent.getRight() == null){
                parent.isLeft() ? parent.parent.deleteLeft() : parent.parent.deleteRight();
                parent = parent.parent;
            }
        }
    }

    while (remainingLength > 0){
        changed.push({el : lastNode, before : lastNode.text});
        if (lastNode.text.match(/[\n]/g))
            specChars--;
        nodeLength = lastNode.getLength();
        parent = lastNode.parent;
        isLeft = lastNode.isLeft();
        if (nodeLength > remainingLength) {
            deletePartFromTextNode(lastNode, 0, remainingLength);
            remainingLength = 0;
        }
        else {
            remainingLength -= nodeLength;
            lastNode.text = null;
            lastNode = nextTextNode(lastNode);
            isLeft ? parent.deleteLeft() : parent.deleteRight();
            while (parent.getLeft() == null && parent.getRight() == null){
                parent.isLeft() ? parent.parent.deleteLeft() : parent.parent.deleteRight();
                parent = parent.parent;
            }
        }
    }

    return changed;
}

function deletePartFromTextNode(node, pos, length){
    let split = splitString(node.text, pos);
    let remaining = splitString(split[1], length);
    node.setText(split[0] + remaining[1]);
}

function concat(left, right) {
    let node = new TreeNode();
    node.setLeft(left);
    node.setRight(right);

    return node;
}

function getAffectedNode(start, pos) {
    let posChecked = 0;
    let node = start;
    while (true) {
        if (node.text)
            break;
        else if (pos === 0) {
            while (node.text == null)
                node = node.getLeft() ?? node.getRight();
        }
        else if (pos === ropeRoot.length){
            while (node.text == null)
                node = node.getRight() ?? node.getLeft();
            posChecked = ropeRoot.length;
        }
        else if (node.getLeft() && (posChecked + node.getLeft().getLength()) >= pos) {
            node = node.getLeft();
        }
        else if (node.getRight() && (posChecked + node.getRight().getLength() + (node.getLeft() ? node.getLeft().getLength() : 0)) >= pos) {
            posChecked += node.getLeft() ? node.getLeft().getLength() : 0;
            node = node.getRight();
        }
        else
            break;
    }
    return [node, posChecked];
}

function nextTextNode(start){
    let node = start;
    let parent = start.parent;

    while (parent){
        if (parent.getRight() === node || parent.getRight() == null){
            node = parent;
            parent = node.parent;
        }
        else
            break;
    }

    if (parent) {
        node = parent.getRight();
        while (node && node.text == null)
            if (node.getLeft())
                node = node.getLeft();
            else
                node = node.getRight();
    }
    else
        node = null

    return node;
}

function prevTextNode(start){
    let node = start;
    let parent = start.parent;

    while (parent){
        if (node.isLeft() || parent.getLeft() == null){
            node = parent;
            parent = node.parent;
        }
        else
            break;
    }

    if (parent) {
        node = parent.getLeft();
        while (node && node.text == null)
            if (node.getRight())
                node = node.getRight();
            else
                node = node.getLeft();
    }
    else
        node = null;

    return node;
}

function getLetterCount(){
    return ropeRoot.getLength() - specChars;
}

function getFullString(start){
    let c = start;
    let toCheck = [];
    while (c != null) {
        toCheck.push(c);
        c = c.getLeft();
    }

    let res = '';
    while (toCheck.length > 0) {
        let node = toCheck.pop();
        if (node.getText()) {
            res = res.concat(node.getText());
        } else {
            let right = node.getRight();
            if (right != null) {
                toCheck.push(right);
                let cleft = right.getLeft();
                while (cleft != null){
                    toCheck.push(cleft);
                    cleft = cleft.getLeft();
                }
            }
        }
    }
    return res;
}

function splitString(str, index) {
    return [str.slice(0, index), str.slice(index)];
}

function getOffset(node){
    let res = 0;
    let curNode = node.parent;
    let prevNode = node;
    while (curNode){
        if (curNode.getLeft() && curNode.getLeft() !== prevNode)
            res += curNode.getLeft().getLength();
        prevNode = curNode;
        curNode = curNode.parent;
    }

    return res;
}

File: src/main/java/ru/saltykov/diploma/editing/DataRope.java
package ru.saltykov.diploma.editing;

import lombok.Getter;
import lombok.Setter;
import org.springframework.data.util.Pair;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class DataRope {
    @Getter
    private TreeNode ropeRoot = new TreeNode();
    @Getter
    @Setter
    public static class TreeNode {
        private TreeNode parent;
        private TreeNode left;
        private TreeNode right;
        private String text;
        private String style;
        private Integer length;

        TreeNode(TreeNode parent, TreeNode left, TreeNode right, String text, String style){
            this.parent = parent;
            this.left = left;
            this.right = right;
            this.text = text;
            this.style = style;
            this.length = text != null ? text.length() : 0;
        }
        TreeNode(){
            this.length = 0;
        }
        public void updateLength(){
            if (this.text == null)
                this.length = (this.getLeft() != null ? this.getLeft().getLength() : 0) + (this.getRight() != null ? this.getRight().getLength() : 0);
            else
                this.length = this.text.length();
            if (this.parent != null)
                this.parent.updateLength();
        }
        public void setLeft(TreeNode left){
            this.left = left;
            if (left != null) this.left.setParent(this);
            this.updateLength();
        }
        public void setRight(TreeNode right){
            this.right = right;
            if (right != null) this.right.setParent(this);
            this.updateLength();
        }
        public void setTest(String text){
            this.text = text;
            this.updateLength();
        }
        public void deleteLeft(){
            if (this.getLeft() != null) {
                this.getLeft().deleteLeft();
                this.getLeft().deleteRight();
                this.getLeft().text = null;
                this.setLeft(null);
            }
            this.updateLength();
        }
        public void deleteRight(){
            if (this.getRight() != null) {
                this.getRight().deleteLeft();
                this.getRight().deleteRight();
                this.getRight().text = null;
                this.setRight(null);
            }
            this.updateLength();
        }
        public boolean isLeft(){
            if (this.getParent() == null)
                return false;
            else
                return Objects.equals(this.getParent().getLeft(), this);
        }
        public TreeNode nextTextNode(){
            if (this.text != null) return DataRope.nextTextNode(this);
            else return DataRope.leftmostChild(this);
        }
        public TreeNode prevTextNode(){
            if (this.text != null) return DataRope.prevTextNode(this);
            else return DataRope.leftmostChild(this);
        }
    }

    public static TreeNode leftmostChild(TreeNode node){
        TreeNode el = node;
        while (el != null && el.getText() == null) {
            if (el.getLeft() != null)
                el = el.getLeft();
            else
                el = el.getRight();
        }
        return el;
    }

    public static TreeNode rightmostChild(TreeNode node){
        TreeNode el = node;
        while (el.getText() == null) {
            if (el.getRight() != null)
                el = el.getRight();
            else
                el = el.getLeft();
        }
        return el;
    }

    public Pair<List<TreeNode>, List<TreeNode>> ropeInsertText(String text, String style, Integer pos){
        TreeNode newNode = new TreeNode(null, null, null, text, style);
        List<TreeNode> added = new ArrayList<>();
        List<TreeNode> removed = new ArrayList<>();
        if (pos == 0 || pos.equals(ropeRoot.getLength())){
            if (ropeRoot.getLeft() != null || ropeRoot.getRight() != null || ropeRoot.getText() != null){
                if (pos == 0){
                    if (ropeRoot.getLeft() == null)
                        ropeRoot.setLeft(newNode);
                    else if (ropeRoot.getRight() == null){
                        ropeRoot.setRight(ropeRoot.getLeft());
                        ropeRoot.setLeft(newNode);
                    }
                    else
                        ropeRoot = concat(newNode, ropeRoot);
                }
                else {
                    if (ropeRoot.getRight() == null)
                        ropeRoot.setRight(newNode);
                    else if (ropeRoot.getLeft() == null) {
                        ropeRoot.setLeft(ropeRoot.getRight());
                        ropeRoot.setRight(newNode);
                    } else
                        ropeRoot = concat(ropeRoot, newNode);
                }
            }
            else
                ropeRoot.setRight(newNode);
            added.add(newNode);
            return Pair.of(added, removed);
        }

        Pair<TreeNode,Integer> toChange = getAffectedNode(ropeRoot, pos);
        TreeNode affectedNode = toChange.getFirst();
        int remainingPos = pos - toChange.getSecond();

        boolean isLeft = affectedNode.isLeft();

        if (remainingPos == 0) {
            if (isLeft) {
                if (affectedNode.parent.getRight() == null) {
                    affectedNode.parent.setRight(affectedNode);
                    affectedNode.parent.setLeft(newNode);
                } else {
                    affectedNode.parent.setLeft(concat(newNode, affectedNode));
                }
            } else {
                if (affectedNode.parent.getLeft() == null) {
                    affectedNode.parent.setLeft(newNode);
                } else {
                    affectedNode.parent.setRight(concat(newNode, affectedNode));
                }
            }
            added.add(newNode);
        } else if (remainingPos == affectedNode.length) {
            if (isLeft) {
                if (affectedNode.parent.getRight() == null) {
                    affectedNode.parent.setRight(newNode);
                } else {
                    affectedNode.parent.setLeft(concat(affectedNode, newNode));
                }
            } else {
                if (affectedNode.parent.getLeft() == null) {
                    affectedNode.parent.setLeft(affectedNode);
                    affectedNode.parent.setRight(newNode);
                } else {
                    affectedNode.parent.setRight(concat(affectedNode, newNode));
                }
            }
            added.add(newNode);
        } else {
            Pair<String, String> split = splitString(affectedNode.text, remainingPos);
            String first = split.getFirst();
            String second = split.getSecond();
            TreeNode node1 = new TreeNode(null, null, null, first, affectedNode.style);
            added.add(node1);
            node1 = concat(node1, newNode);
            TreeNode node2 = new TreeNode(null, null, null, second, affectedNode.style);
            node1 = concat(node1, node2);
            if (isLeft)
                affectedNode.parent.setLeft(node1);
            else
                affectedNode.parent.setRight(node1);
            added.add(newNode);
            added.add(node2);
            removed.add(affectedNode);
        }

        return Pair.of(added, removed);
    }

    public List<Pair<TreeNode, String>> ropeDeleteText(int pos, int length){
        Pair<TreeNode, Integer> start = getAffectedNode(ropeRoot, pos + 1);
        TreeNode affected = start.getFirst();
        int remainingPos = pos - start.getSecond();
        if (remainingPos < 0)
            remainingPos = Math.abs(pos - start.getSecond());

        int remainingLength = length;

        TreeNode lastNode = affected;
        int nodeLength = lastNode.getLength();
        TreeNode parent = lastNode.getParent();
        boolean isLeft = lastNode.isLeft();

        List<Pair<TreeNode, String>> changed = new ArrayList<>();
        if (length == ropeRoot.getLength()){
            TreeNode el = leftmostChild(ropeRoot);
            while (el != null){
                changed.add(Pair.of(el, el.getText()));
                el = el.nextTextNode();
            }
            ropeRoot.deleteLeft();
            ropeRoot.deleteRight();
            return changed;
        }

        if (remainingPos > 0){
            int toDelete =  Math.max(0, Math.min(nodeLength - remainingPos, remainingLength));
            remainingLength -= toDelete;
            changed.add(Pair.of(lastNode, lastNode.getText()));
            deletePartFromTextNode(lastNode, remainingPos, toDelete);
            lastNode = lastNode.nextTextNode();
        }
        else {
            changed.add(Pair.of(lastNode, lastNode.getText()));
            if (nodeLength > remainingLength) {
                deletePartFromTextNode(lastNode, 0, remainingLength);
                remainingLength = 0;
            }
            else {
                lastNode.text = null;
                remainingLength -= nodeLength;
                lastNode = lastNode.nextTextNode();
                if (isLeft) parent.deleteLeft();
                else parent.deleteRight();

                while (parent.getLeft() == null && parent.getRight() == null){
                    if (parent.isLeft())
                        parent.parent.deleteLeft();
                    else parent.parent.deleteRight();
                    parent = parent.getParent();
                }
            }
        }

        while (remainingLength > 0){
            changed.add(Pair.of(lastNode, lastNode.getText()));
            nodeLength = lastNode.getLength();
            parent = lastNode.parent;
            isLeft = lastNode.isLeft();
            if (nodeLength > remainingLength) {
                deletePartFromTextNode(lastNode, 0, remainingLength);
                remainingLength = 0;
            }
            else {
                remainingLength -= nodeLength;
                lastNode.text = null;
                lastNode = lastNode.nextTextNode();
                if (isLeft)
                    parent.deleteLeft();
                else
                    parent.deleteRight();
                while (parent.getLeft() == null && parent.getRight() == null){
                    if (parent.isLeft())
                        parent.parent.deleteLeft();
                    else
                        parent.parent.deleteRight();
                    parent = parent.parent;
                }
            }
        }

        return changed;
    }

    public void deletePartFromTextNode(TreeNode node, int pos, int length){
        Pair<String, String> split = splitString(node.getText(), pos);
        Pair<String, String> remaining = splitString(split.getSecond(), length);
        node.setText(split.getFirst() + remaining.getSecond());
    }

    public TreeNode concat(TreeNode left, TreeNode right) {
        TreeNode node = new TreeNode();
        node.setLeft(left);
        node.setRight(right);

        return node;
    }

    public Pair<TreeNode, Integer> getAffectedNode(TreeNode start, int pos) {
        int posChecked = 0;
        TreeNode node = start;
        while (true) {
            if (node.getText() != null)
                break;
            else if (pos == 0) {
                while (node.getText() == null)
                    node = node.getLeft() != null ? node.getLeft() : node.getRight();
            }
            else if (pos == ropeRoot.getLength()){
                while (node.getText() == null)
                    node = node.getRight() != null ? node.getRight() : node.getLeft();
                posChecked = ropeRoot.getLength();
            }
            else if (node.getLeft() != null && (posChecked + node.getLeft().getLength()) >= pos) {
                node = node.getLeft();
            }
            else if (node.getRight() != null && (posChecked + node.getRight().getLength() + (node.getLeft() != null ? node.getLeft().getLength() : 0)) >= pos) {
                posChecked += node.getLeft() != null ? node.getLeft().getLength() : 0;
                node = node.getRight();
            }
            else
                break;
        }
        return Pair.of(node, posChecked);
    }

    public static TreeNode nextTextNode(TreeNode start){
        TreeNode node = start;
        TreeNode parent = start.getParent();

        while (parent != null){
            if (parent.getRight() == null || parent.getRight().equals(node)){
                node = parent;
                parent = node.getParent();
            }
            else
                break;
        }

        if (parent != null) {
            node = parent.getRight();
            while (node != null && node.getText() == null)
                if (node.getLeft() != null)
                    node = node.getLeft();
                else
                    node = node.getRight();
        }
        else
            node = null;

        return node;
    }

    public static TreeNode prevTextNode(TreeNode start){
        TreeNode node = start;
        TreeNode parent = start.getParent();

        while (parent != null){
            if (node.isLeft() || parent.getLeft() == null){
                node = parent;
                parent = node.getParent();
            }
            else
                break;
        }

        if (parent != null) {
            node = parent.getLeft();
            while (node != null && node.getText() == null)
                if (node.getRight() != null)
                    node = node.getRight();
                else
                    node = node.getLeft();
        }
        else
            node = null;

        return node;
    }

    public Pair<String, String> splitString(String str, int index) {
        return Pair.of(str.substring(0, index), str.substring(index));
    }
}
